File: frontend/src/app/models/gacha.ts
export enum RarityTier {
  N = 'N',
  R = 'R',
  SR = 'SR',
  SSR = 'SSR',
  UR = 'UR',
}

export enum GachaPullMode {
  SINGLE = 'single',
  TEN_PULL = 'ten_pull',
  DAILY = 'daily',
}

export interface GachaPullRequest {
  mode?: GachaPullMode;
  pool?: string | null;
}

export interface GachaPullItemRead {
  id: string;
  media_id: string;
  rarity_tier: RarityTier;
  rarity_score: number;
  was_duplicate: boolean;
  upgrade_material_granted: number;
  position: number;
  collection_item_id: string | null;
}

export interface GachaPullRead {
  id: string;
  user_id: string;
  mode: GachaPullMode;
  pool: string | null;
  currency_spent: number;
  currency_balance: number | null;
  created_at: string;
  items: GachaPullItemRead[];
}

export interface RaritySnapshotRead {
  media_id: string;
  rarity_score: number;
  rarity_tier: RarityTier;
  component_scores: Record<string, number>;
  score_version: string;
  previous_tier: RarityTier | null;
  below_threshold_count: number;
  calculated_at: string;
}

export interface GachaStatsResponse {
  total_rarity_snapshots: number;
  tier_counts: Partial<Record<RarityTier, number>>;
  collection_count: number;
  duplicate_copies: number;
  currency_balance: number;
  daily_claim_available: boolean;
  next_daily_claim_at: string | null;
}

export interface RarityRecalculationResponse {
  recalculated: number;
  score_version: string;
  tier_counts: Partial<Record<RarityTier, number>>;
}

export interface GachaCurrencyBalanceRead {
  user_id: string;
  balance: number;
  total_claimed: number;
  total_spent: number;
  last_daily_claimed_on: string | null;
  daily_claim_amount: number;
  daily_claim_available: boolean;
  next_daily_claim_at: string | null;
}

export interface GachaDailyClaimResponse {
  claimed: number;
  balance: number;
  daily_claim_available: boolean;
  next_daily_claim_at: string;
}

File: backend/app/schemas/gacha.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field

from backend.app.models.gacha import GachaPullMode, RarityTier


class GachaPullRequest(BaseModel):
    mode: GachaPullMode = Field(default=GachaPullMode.single)
    pool: str | None = Field(default=None, max_length=64)


class GachaPullItemRead(BaseModel):
    id: uuid.UUID
    media_id: uuid.UUID
    rarity_tier: RarityTier
    rarity_score: float
    was_duplicate: bool
    upgrade_material_granted: int
    position: int
    collection_item_id: uuid.UUID | None = None


class GachaPullRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    mode: GachaPullMode
    pool: str | None = None
    currency_spent: int = 0
    currency_balance: int | None = None
    created_at: datetime
    items: list[GachaPullItemRead]


class RaritySnapshotRead(BaseModel):
    media_id: uuid.UUID
    rarity_score: float
    rarity_tier: RarityTier
    component_scores: dict
    score_version: str
    previous_tier: RarityTier | None = None
    below_threshold_count: int
    calculated_at: datetime

    model_config = {"from_attributes": True}


class GachaStatsResponse(BaseModel):
    total_rarity_snapshots: int
    tier_counts: dict[RarityTier, int]
    collection_count: int
    duplicate_copies: int
    currency_balance: int
    daily_claim_available: bool
    next_daily_claim_at: datetime | None = None


class RarityRecalculationResponse(BaseModel):
    recalculated: int
    score_version: str
    tier_counts: dict[RarityTier, int]


class GachaCurrencyBalanceRead(BaseModel):
    user_id: uuid.UUID
    balance: int
    total_claimed: int
    total_spent: int
    last_daily_claimed_on: date | None = None
    daily_claim_amount: int
    daily_claim_available: bool
    next_daily_claim_at: datetime | None = None


class GachaDailyClaimResponse(BaseModel):
    claimed: int
    balance: int
    daily_claim_available: bool
    next_daily_claim_at: datetime
